### PR TITLE
[TAG] Create GRPC client/server

### PR DESCRIPTION
**Pull request by jakule**
This PR replaces Teleport <- REST -> TAG communication with GRPC. It  adds the new GRPC service with Go and Python implementation as we need Go client and Python server. It also changes the handler to use new Go GRPC client.

### CODE REVIEW

File: buf-go.gen.yaml [hunk 1]
```diff
@@ -22,3 +22,4 @@ plugins:
       - module=github.com/gravitational/teleport
       # needed by teleport/lib/teleterm/v1/usage_events.proto
       - Mprehog/v1alpha/connect.proto=github.com/gravitational/teleport/gen/proto/go/prehog/v1alpha;prehogv1alpha
+      - Maccessgraph/v1/query.proto=github.com/gravitational/teleport/gen/proto/go/accessgraph/v1;accessgraphv1
```
Comment (codingllama) [inline buf-go.gen.yaml:25]: Is the comment above true for this line? ("needed by ...")

File: build.assets/genproto.sh [hunk 2]
```diff
@@ -74,6 +74,19 @@ main() {
   [[ $skip_js -eq 0 ]] && echoed buf generate --template=buf-js.gen.yaml \
     --path=proto/prehog/ \
     --path=proto/teleport/lib/teleterm/
+
+  echoed buf generate --template=buf-python.gen.yaml \
```
Comment (codingllama) [inline build.assets/genproto.sh:78]:
```suggestion
  # Generate Python protos.
  echoed buf generate --template=buf-python.gen.yaml \
```

File: build.assets/genproto.sh [hunk 3]
```diff
@@ -74,6 +74,19 @@ main() {
...
+      --path=proto/accessgraph/
+
+  # Add __init__.py file to the root directory with generated Python packages.
+  cat <<EOF > gen/proto/py/__init__.py
+# This file is generated by build-assets/genproto.sh.
+# DO NOT EDIT THIS FILE BY HAND.
```
Comment (codingllama) [inline build.assets/genproto.sh:84]: Maybe, so it matches pb.go headers?

```suggestion
# Code generated by build-assets/genproto.sh. DO NOT EDIT.

```

File: lib/service/service.go [hunk 4]
```diff
@@ -4008,9 +4008,13 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 			traceClt = clt
 		}
 
-		var accessGraphAddr string
+		var accessGraphAddr utils.NetAddr
```
Comment (codingllama) [inline lib/service/service.go:4011]: Could we keep this focused on python codegen and move other changes to a separate PR?

File: proto/accessgraph/v1/query.proto
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:1]: Call the file query_service.proto, so it's clear that it holds a service?

File: proto/accessgraph/v1/query.proto [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+option go_package = "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1;accessgraphv1";
+
+// QueryRequest is a request to query the access graph.
+message QueryRequest {
+  string query = 1;
+}
+
+// Node is a node in the access graph.
+message Node {
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:27]: I would typically move important entities, like Node and Edge, to their own files.

File: proto/accessgraph/v1/query.proto [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  string type = 3;
+}
+
+// QueryResponse is a response to a query.
+message QueryResponse {
+  repeated Node nodes = 1;
+  repeated Edge edges = 2;
+}
+
+// QueryService is a service for querying the access graph.
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:49]: Is there anything more interesting we could say about the service?

File: proto/accessgraph/v1/query.proto [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+service QueryService {
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:50]: nit: Move service to the top of the file? The service is the interesting part, the requests and responses are mainly boilerplate.

File: proto/accessgraph/v1/query.proto [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  // Query queries the access graph.
+  rpc Query(QueryRequest) returns (QueryResponse);
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:52]: "QueryService" and "Query" are rather generic names. Consider "AccessGraphService" and "ListNodes", for example, as more descriptive names.

As all things, this is likely to evolve over time, getting new methods and responsibilities. When that happens, the overly generic names will get more and more ambiguious.

I'd also strongly suggest a resource-oriented design, like this: https://cloud.google.com/apis/design/resources.

File: proto/accessgraph/v1/query.proto [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+syntax = "proto3";
+
+package accessgraph.v1;
+
+option go_package = "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1;accessgraphv1";
+
+// QueryRequest is a request to query the access graph.
+message QueryRequest {
+  string query = 1;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:23]: What is a query string? An example would go a long way.

File: proto/accessgraph/v1/query.proto [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// Node is a node in the access graph.
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:26]: What can a node represent?

File: proto/accessgraph/v1/query.proto [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+message Node {
+  string id = 1;
+  string kind = 2;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:29]: What is the kind for nodes?

File: proto/accessgraph/v1/query.proto [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  string sub_kind = 3;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:30]: Are sub_kinds used for nodes? What are the possible values?

File: proto/accessgraph/v1/query.proto [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  string name = 4;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:31]: Is the node-name user supplied? Is it unique? What are some examples?

File: proto/accessgraph/v1/query.proto [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  map<string, string> labels = 5;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:32]: What kinds of labels should we expect to find here?

File: proto/accessgraph/v1/query.proto [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+  string hostname = 6;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:33]: Do all nodes have a hostname?

File: proto/accessgraph/v1/query.proto [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// Edge is an edge in the access graph.
+message Edge {
+  string from = 1;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:38]: What do from and to contain?

File: proto/accessgraph/v1/query.proto [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  string to = 2;
+  string type = 3;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:40]: What types of edges are available?

Why not an enum?

File: proto/accessgraph/v1/query.proto [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// QueryResponse is a response to a query.
+message QueryResponse {
+  repeated Node nodes = 1;
```
Comment (codingllama) [inline proto/accessgraph/v1/query.proto:45]: It's difficult to understand how a response could hold multiple nodes and edges without knowing what a query might be.

**Review by codingllama — COMMENTED** [review]
I've got various design questions, apologies if these are already covered somewhere else. (In any case, the questions remain.)

As mentioned in other comment, I would keep only protos and protogen here and move code changes to another PR.

**Comment by codingllama** [comment]
I just noticed this is targeted for dev-tag - feel free to ignore me if this is not the time to address those comments.

File: proto/accessgraph/v1/query.proto (hunk 6 above)
Comment (justinas) [reply proto/accessgraph/v1/query.proto:49]: I would add: this is currently used only for presentation purposes (e.g. Web UI). Once TAG is involved in the access checks, this same API may or may not be used.

File: proto/accessgraph/v1/query.proto (hunk 15 above)
Comment (justinas) [reply proto/accessgraph/v1/query.proto:33]: ➕ on questions on purpose of `labels` and `hostname`. If this is specific to resources like SSH hosts (as opposed to users, actions, etc.), we may want to start using something like `oneof` to carry additional data specific to the kind of node.

**Comment by justinas** [comment]
One big concern is backward/forward compatibility between TAG and teleport(.e). Do we have any ideas for this? Do we want to proactively make everything versioned a la QueryResponseV1?

File: proto/accessgraph/v1/query.proto (hunk 18 above)
Comment (justinas) [reply proto/accessgraph/v1/query.proto:45]: The query language is SQL-ish, and the underlying response is SQL-ish as well (tabular). However, the UI (which is the ultimate client of QueryService) does not care about the tabular format in particular: it wants to render the selected (according to the query) subgraph of the entire access graph. Hence, nodes and edges are enough. Not sure why the query would need to be returned as well?

Comment (codingllama) [reply proto/accessgraph/v1/query.proto:45]: Just to clarify I'm not asking for the query to be returned, I'm just saying that it's difficult for a reader to understand the response without any explanation about what the input is.

**Comment by jakule** [comment]
> One big concern is backward/forward compatibility between TAG and teleport(.e). Do we have any ideas for this? Do we want to proactively make everything versioned a la QueryResponseV1?

That is also my concern, but I can see two solutions
1. We go with the current solution where every fields is the part of proto (we could use `oneof` for different resource types) This approach requires Teleport to be updated when a new field is added, but also give us some validation of what data are returned.
2. We send all data as `map<string, string>` and not worry about the proto definition (only TAG UI would need to be updated, but we can skip the Teleport), but that would make hard to validate the correctness IMO. It would also require more implementation in the UI as only some field would be populated ex. `host_id` would be only set when the `subkind == node`. 

@tigrato WDYT

**Comment by tigrato** [comment]
> > One big concern is backward/forward compatibility between TAG and teleport(.e). Do we have any ideas for this? Do we want to proactively make everything versioned a la QueryResponseV1?
> 
> That is also my concern, but I can see two solutions
> 
>     1. We go with the current solution where every fields is the part of proto (we could use `oneof` for different resource types) This approach requires Teleport to be updated when a new field is added, but also give us some validation of what data are returned.
> 
>     2. We send all data as `map<string, string>` and not worry about the proto definition (only TAG UI would need to be updated, but we can skip the Teleport), but that would make hard to validate the correctness IMO. It would also require more implementation in the UI as only some field would be populated ex. `host_id` would be only set when the `subkind == node`.
> 
> 
> @tigrato WDYT

I actually would prefer `protobuf.Struct` wrapper - it works as a `map[string]any` 

I already used it here: https://github.com/gravitational/teleport/blob/04774100cc05de46574cfea39085a930f62e5591/api/proto/teleport/auditlog/v1/auditlog.proto#L93

During the development phase where we need to deploy daily it's the best approach since it doesn't require updating Teleport and we can send any structured data from TAG service  and marshal it directly in the API. Teleport works as a simple proxy. Once we finish this active development we can create better apis using oneOfs.
I am a bit concerned about having to update teleport to add new data required in the API

**Comment by jakule** [comment]
> > > One big concern is backward/forward compatibility between TAG and teleport(.e). Do we have any ideas for this? Do we want to proactively make everything versioned a la QueryResponseV1?
> > 
> > 
> > That is also my concern, but I can see two solutions
> > ```
> > 1. We go with the current solution where every fields is the part of proto (we could use `oneof` for different resource types) This approach requires Teleport to be updated when a new field is added, but also give us some validation of what data are returned.
> > 
> > 2. We send all data as `map<string, string>` and not worry about the proto definition (only TAG UI would need to be updated, but we can skip the Teleport), but that would make hard to validate the correctness IMO. It would also require more implementation in the UI as only some field would be populated ex. `host_id` would be only set when the `subkind == node`.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @tigrato WDYT
> 
> I actually would prefer `protobuf.Struct` wrapper - it works as a `map[string]any`
> 
> I already used it here:
> 
> https://github.com/gravitational/teleport/blob/04774100cc05de46574cfea39085a930f62e5591/api/proto/teleport/auditlog/v1/auditlog.proto#L93
> 
> During the development phase where we need to deploy daily it's the best approach since it doesn't require updating Teleport and we can send any structured data from TAG service and marshal it directly in the API. Teleport works as a simple proxy. Once we finish this active development we can create better apis using oneOfs. I am a bit concerned about having to update teleport to add new data required in the API

@tigrato @justinas Let me ask a different question. Do we think we will need to keep adding many new fields to the UI? I was thinking about the problem, but I think we deviated a bit from the initial problem. In principle, there is no reason why proxy should know proto definition as this makes the whole process more complicated, but the GRPC/Proto library doesn't allow us to proxy messages without the proxy knowing the definition. Workarounds that I found is to use this JS library https://github.com/protobufjs/protobuf.js to unmarshal messages in the UI. We could easily share proto-definition between TAG and WebUI, so this shouldn't be a problem. 
The second option would be to send JSON as a string over the wire, literally use
```
message QueryResponse {
    string data = 1;
}
```
or any combination of `map<string, string>`/`protobuf.Struct` that we discussed above. 
My personal complaint about using `map` or `protobuf.Struct` is that we don't know what fields are in the message (as there is no proto to review/reason about) and without having any object definition, we may send some unwanted data to the UI.

Note: We can rename the version to `v1alpha` and change the API in `v1` if we're not happy. Let's decide on something and change it later rather than keep debating about what is the best.

**Comment by tigrato** [comment]
> Let me ask a different question. Do we think we will need to keep adding many new fields to the UI? I was thinking about the problem, but I think we deviated a bit from the initial problem. In principle, there is no reason why proxy should know proto definition as this makes the whole process more complicated, but the GRPC/Proto library doesn't allow us to proxy messages without the proxy knowing the definition.

I don't think we are going to add new fields  that often, my concern is more about delayed upgrades where some tenants don't receive the updates because they are old and somehow TAG has to know it

File: proto/accessgraph/v1/query.proto (hunk 10 above)
Comment (jakule) [reply proto/accessgraph/v1/query.proto:26]: Node can represent anything. I'm not sure how much domain specific knowledge should proto carry compare to docs TBH.

File: proto/accessgraph/v1/query.proto (hunk 11 above)
Comment (jakule) [reply proto/accessgraph/v1/query.proto:29]: TBD Currently we have Node/ssh, but this will change soon.

File: proto/accessgraph/v1/query.proto (hunk 10 above)
Comment (codingllama) [reply proto/accessgraph/v1/query.proto]: At the bare minimum enough for people looking at the protos do have an idea of what it can do, but if you ask me this is the place for dev-focused docs.
